File: Applications/ZylkerKart_Ecommerce/services/payment-service/chaos-sdk/src/site24x7_chaos/fault/resource.py
"""Resource / JVM-equivalent fault injector — 6 fault types.

All faults run in daemon threads with a bounded duration so they
self-clean and cannot outlive the process.

Fault types handled
-------------------
1. ``thread_pool_exhaustion`` — Spawn busy-sleeping threads.
2. ``memory_pressure``        — Allocate large bytearrays on the heap.
3. ``cpu_burn``               — Tight math loops across multiple threads.
4. ``gc_pressure``            — Rapid short-lived allocations to stress the GC.
5. ``thread_deadlock``        — Two threads that deadlock on ``threading.Lock``.
6. ``disk_fill``              — Write temp files to consume disk space.

Trigger model
-------------
Unlike HTTP/JDBC/Redis faults, resource faults are **not** triggered per-request.
Instead, the :class:`ResourceFaultInjector` registers as a config-update listener
on the :class:`ChaosEngine`.  When the agent pushes a new config, the injector
checks for active resource rules and fires the first one (only one resource fault
can be active at a time to prevent cascading issues).
"""

from __future__ import annotations

import gc
import logging
import math
import os
import random
import tempfile
import threading
import time
from pathlib import Path
from typing import TYPE_CHECKING, List

from ..models import FaultRuleConfig

if TYPE_CHECKING:
    from ..engine import ChaosEngine

logger = logging.getLogger("site24x7_chaos.fault.resource")

_RESOURCE_FAULT_TYPES = frozenset(
    {
        "thread_pool_exhaustion",
        "memory_pressure",
        "cpu_burn",
        "gc_pressure",
        "thread_deadlock",
        "disk_fill",
    }
)


class ResourceFaultInjector:
    """Registers as a config-update listener and fires resource faults.

    Usage::

        injector = ResourceFaultInjector(engine)
        injector.install()   # registers the listener
    """

    def __init__(self, engine: ChaosEngine) -> None:
        self._engine = engine
        self._active = threading.Event()

    def install(self) -> None:
        """Register as a config-update listener on the engine."""
        self._engine.add_config_update_listener(self._on_config_update)
        logger.info("ResourceFaultInjector installed")

    # ------------------------------------------------------------------
    # Listener
    # ------------------------------------------------------------------

    def _on_config_update(self) -> None:
        self.evaluate_and_apply()

    def evaluate_and_apply(self) -> None:
        """Check active rules for resource faults and fire the first match."""
        if not self._engine.enabled:
            return

        if self._active.is_set():
            logger.debug("Resource fault already active, skipping")
            return

        rules: List[FaultRuleConfig] = [
            r
            for r in self._engine.active_rules
            if r.fault_type in _RESOURCE_FAULT_TYPES and self._engine.should_fire(r)
        ]

        if not rules:
            return

        rule = rules[0]
        ft = rule.fault_type
        if ft == "thread_pool_exhaustion":
            self._apply_thread_pool_exhaustion(rule)
        elif ft == "memory_pressure":
            self._apply_memory_pressure(rule)
        elif ft == "cpu_burn":
            self._apply_cpu_burn(rule)
        elif ft == "gc_pressure":
            self._apply_gc_pressure(rule)
        elif ft == "thread_deadlock":
            self._apply_thread_deadlock(rule)
        elif ft == "disk_fill":
            self._apply_disk_fill(rule)
        else:
            logger.warning("Unknown resource fault type: %s", ft)

    # ------------------------------------------------------------------
    # 1. Thread pool exhaustion
    # ------------------------------------------------------------------

    def _apply_thread_pool_exhaustion(self, rule: FaultRuleConfig) -> None:
        thread_count = _clamp(rule.get_config_int("thread_count", 10), 1, 50)
        duration_ms = _clamp(rule.get_config_int("duration_ms", 30000), 1000, 60000)

        logger.info(
            "Injecting thread pool exhaustion: %d threads for %dms",
            thread_count,
            duration_ms,
        )
        self._active.set()

        def orchestrator() -> None:
            barrier = threading.Barrier(thread_count + 1, timeout=duration_ms / 1000.0 + 5)
            workers: list[threading.Thread] = []

            def worker() -> None:
                try:
                    time.sleep(duration_ms / 1000.0)
                finally:
                    try:
                        barrier.wait()
                    except threading.BrokenBarrierError:
                        pass

            try:
                for i in range(thread_count):
                    t = threading.Thread(
                        target=worker,
                        name=f"chaos-thread-exhaust-{i}",
                        daemon=True,
                    )
                    workers.append(t)
                    t.start()
                try:
                    barrier.wait()
                except threading.BrokenBarrierError:
                    pass
                logger.info("Thread pool exhaustion fault completed")
            finally:
                self._active.clear()

        threading.Thread(
            target=orchestrator,
            name="chaos-thread-exhaust-orchestrator",
            daemon=True,
        ).start()

    # ------------------------------------------------------------------
    # 2. Memory pressure
    # ------------------------------------------------------------------

    def _apply_memory_pressure(self, rule: FaultRuleConfig) -> None:
        allocation_mb = _clamp(rule.get_config_int("allocation_mb", 64), 1, 512)
        duration_ms = _clamp(rule.get_config_int("duration_ms", 30000), 1000, 60000)

        logger.info(
            "Injecting memory pressure: %dMB for %dms",
            allocation_mb,
            duration_ms,
        )
        self._active.set()

        def mem_worker() -> None:
            blocks: list[bytearray] | None = None
            try:
                blocks = []
                for i in range(allocation_mb):
                    block = bytearray(1024 * 1024)  # 1 MB
                    # Touch the memory at page boundaries to ensure real allocation
                    for offset in range(0, len(block), 4096):
                        block[offset] = i & 0xFF
                    blocks.append(block)
                logger.debug("Memory pressure: allocated %dMB", allocation_mb)
                time.sleep(duration_ms / 1000.0)
            except MemoryError:
                logger.warning(
                    "Memory pressure: MemoryError during allocation, holding partial"
                )
                time.sleep(duration_ms / 1000.0)
            finally:
                del blocks
                gc.collect()
                self._active.clear()
                logger.info("Memory pressure fault completed, memory released")

        threading.Thread(
            target=mem_worker, name="chaos-memory-pressure", daemon=True
        ).start()

    # ------------------------------------------------------------------
    # 3. CPU burn
    # ------------------------------------------------------------------

    def _apply_cpu_burn(self, rule: FaultRuleConfig) -> None:
        thread_count = _clamp(rule.get_config_int("thread_count", 2), 1, 8)
        duration_ms = _clamp(rule.get_config_int("duration_ms", 30000), 1000, 60000)

        logger.info(
            "Injecting CPU burn: %d threads for %dms", thread_count, duration_ms
        )
        self._active.set()

        end_time = time.monotonic() + duration_ms / 1000.0
        barrier = threading.Barrier(thread_count + 1, timeout=duration_ms / 1000.0 + 5)

        def burner() -> None:
            try:
                x = random.random()
                while time.monotonic() < end_time:
                    for _ in range(1000):
                        x = math.sin(x) * math.cos(x)
            finally:
                try:
                    barrier.wait()
                except threading.BrokenBarrierError:
                    pass

        def orchestrator() -> None:
            try:
                for i in range(thread_count):
                    threading.Thread(
                        target=burner,
                        name=f"chaos-cpu-burn-{i}",
                        daemon=True,
                    ).start()
                try:
                    barrier.wait()
                except threading.BrokenBarrierError:
                    pass
                logger.info("CPU burn fault completed")
            finally:
                self._active.clear()

        threading.Thread(
            target=orchestrator,
            name="chaos-cpu-burn-orchestrator",
            daemon=True,
        ).start()

    # ------------------------------------------------------------------
    # 4. GC pressure
    # ------------------------------------------------------------------

    def _apply_gc_pressure(self, rule: FaultRuleConfig) -> None:
        alloc_rate = _clamp(
            rule.get_config_int("allocation_rate_mb_per_sec", 10), 1, 100
        )
        duration_ms = _clamp(rule.get_config_int("duration_ms", 30000), 1000, 60000)

        logger.info(
            "Injecting GC pressure: %dMB/sec for %dms", alloc_rate, duration_ms
        )
        self._active.set()

        def gc_worker() -> None:
            try:
                end_time = time.monotonic() + duration_ms / 1000.0
                sleep_s = max(0.001, 1.0 / alloc_rate)
                while time.monotonic() < end_time:
                    # Allocate 1MB — short-lived so the GC must collect it
                    garbage = bytearray(1024 * 1024)
                    garbage[0] = 1
                    garbage[-1] = 1
                    del garbage
                    time.sleep(sleep_s)
                logger.info("GC pressure fault completed")
            except MemoryError:
                logger.warning("GC pressure: MemoryError, stopping early")
            finally:
                self._active.clear()

        threading.Thread(
            target=gc_worker, name="chaos-gc-pressure", daemon=True
        ).start()

    # ------------------------------------------------------------------
    # 5. Thread deadlock
    # ------------------------------------------------------------------

    def _apply_thread_deadlock(self, rule: FaultRuleConfig) -> None:
        duration_ms = _clamp(rule.get_config_int("duration_ms", 30000), 1000, 60000)

        logger.info("Injecting thread deadlock for %dms", duration_ms)
        self._active.set()

        lock_a = threading.Lock()
        lock_b = threading.Lock()
        stop_event = threading.Event()

        def thread_1() -> None:
            lock_a.acquire()
            try:
                time.sleep(0.05)
                # This will block forever (deadlock) until stop_event
                while not stop_event.is_set():
                    if lock_b.acquire(timeout=0.1):
                        try:
                            time.sleep(duration_ms / 1000.0)
                        finally:
                            lock_b.release()
                        break
            finally:
                lock_a.release()

        def thread_2() -> None:
            lock_b.acquire()
            try:
                time.sleep(0.05)
                while not stop_event.is_set():
                    if lock_a.acquire(timeout=0.1):
                        try:
                            time.sleep(duration_ms / 1000.0)
                        finally:
                            lock_a.release()
                        break
            finally:
                lock_b.release()

        def watchdog() -> None:
            try:
                t1 = threading.Thread(
                    target=thread_1, name="chaos-deadlock-1", daemon=True
                )
                t2 = threading.Thread(
                    target=thread_2, name="chaos-deadlock-2", daemon=True
                )
                t1.start()
                t2.start()
                time.sleep(duration_ms / 1000.0)
            finally:
                stop_event.set()
                self._active.clear()
                logger.info("Thread deadlock fault completed")

        threading.Thread(
            target=watchdog, name="chaos-deadlock-watchdog", daemon=True
        ).start()

    # ------------------------------------------------------------------
    # 6. Disk fill
    # ------------------------------------------------------------------

    def _apply_disk_fill(self, rule: FaultRuleConfig) -> None:
        allocation_mb = _clamp(rule.get_config_int("allocation_mb", 64), 1, 512)
        duration_ms = _clamp(rule.get_config_int("duration_ms", 30000), 1000, 60000)

        logger.info(
            "Injecting disk fill: %dMB for %dms", allocation_mb, duration_ms
        )
        self._active.set()

        def disk_worker() -> None:
            temp_files: list[Path] = []
            try:
                tmp_dir = Path(tempfile.gettempdir())
                block = b"\xAA" * (1024 * 1024)  # 1 MB

                for i in range(allocation_mb):
                    fp = tmp_dir / f"chaos-disk-fill-{i}.tmp"
                    fp.write_bytes(block)
                    temp_files.append(fp)

                logger.debug("Disk fill: wrote %dMB of temp files", allocation_mb)
                time.sleep(duration_ms / 1000.0)
            except OSError as exc:
                logger.warning("Disk fill: error during writing: %s", exc)
            finally:
                for fp in temp_files:
                    try:
                        fp.unlink(missing_ok=True)
                    except OSError:
                        pass
                self._active.clear()
                logger.info(
                    "Disk fill fault completed, %d temp files cleaned up",
                    len(temp_files),
                )

        threading.Thread(
            target=disk_worker, name="chaos-disk-fill", daemon=True
        ).start()


# ------------------------------------------------------------------
# Helpers
# ------------------------------------------------------------------


def _clamp(value: int, lo: int, hi: int) -> int:
    return max(lo, min(hi, value))

File: Applications/ZylkerKart_Ecommerce/services/search-service/chaos-sdk/fault/resource.go
// Site24x7 Labs Chaos SDK for Go — Resource Fault Injector
//
// Unlike HTTP/DB/Redis faults (triggered per-request), resource faults are
// triggered when the config changes. The injector registers as a
// config-update listener on the ChaosEngine.
//
// 6 fault types:
//  1. thread_pool_exhaustion — Leaked goroutines blocked on a channel.
//  2. memory_pressure        — Allocate []byte slices and hold them.
//  3. cpu_burn               — LockOSThread + tight math loops.
//  4. gc_pressure            — Rapid short-lived allocations.
//  5. thread_deadlock        — Goroutines blocked on channel with no writer.
//  6. disk_fill              — Write temp files to consume disk space.
package fault

import (
	"fmt"
	"log"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"sync/atomic"
	"time"

	chaos "github.com/pazhanir/site24x7-labs/sdks/go"
)

// ---------------------------------------------------------------------------
// ResourceFaultInjector
// ---------------------------------------------------------------------------

// ResourceFaultInjector listens for config changes and injects resource
// exhaustion faults. Only one resource fault is active at a time.
type ResourceFaultInjector struct {
	engine *chaos.Engine
	active atomic.Bool
}

// NewResourceFaultInjector creates a new injector. Call Install() to register
// the config-update listener.
func NewResourceFaultInjector(engine *chaos.Engine) *ResourceFaultInjector {
	return &ResourceFaultInjector{engine: engine}
}

// Install registers this injector as a config-update listener on the engine.
// Call once after the engine is started.
func (r *ResourceFaultInjector) Install() {
	r.engine.AddConfigUpdateListener(func() {
		r.evaluateAndApply()
	})
	log.Println("[chaos-sdk] ResourceFaultInjector installed")
}

// evaluateAndApply finds the first matching resource fault rule and applies it.
func (r *ResourceFaultInjector) evaluateAndApply() {
	if !r.engine.Enabled() {
		return
	}
	if r.active.Load() {
		log.Println("[chaos-sdk] Resource fault already active, skipping")
		return
	}

	// Find the first matching resource fault rule
	rules := r.engine.FindMatchingRules("")
	for _, rule := range rules {
		if !chaos.ResourceFaultTypes[rule.FaultType] {
			continue
		}
		if !r.engine.ShouldFire(rule) {
			continue
		}

		switch rule.FaultType {
		case chaos.FaultThreadPoolExhaustion:
			r.applyThreadPoolExhaustion(rule)
		case chaos.FaultMemoryPressure:
			r.applyMemoryPressure(rule)
		case chaos.FaultCPUBurn:
			r.applyCPUBurn(rule)
		case chaos.FaultGCPressure:
			r.applyGCPressure(rule)
		case chaos.FaultThreadDeadlock:
			r.applyThreadDeadlock(rule)
		case chaos.FaultDiskFill:
			r.applyDiskFill(rule)
		default:
			log.Printf("[chaos-sdk] Unknown resource fault type: %s", rule.FaultType)
		}
		return // Only inject one at a time
	}
}

// ---------------------------------------------------------------------------
// 1. Thread pool exhaustion — leaked goroutines
// ---------------------------------------------------------------------------

func (r *ResourceFaultInjector) applyThreadPoolExhaustion(rule chaos.FaultRuleConfig) {
	threadCount := chaos.Clamp(chaos.GetConfigInt(rule, "thread_count", 10), 1, 50)
	durationMs := chaos.Clamp(chaos.GetConfigInt(rule, "duration_ms", 30000), 1000, 60000)

	log.Printf("[chaos-sdk] Injecting goroutine leak: %d goroutines for %dms", threadCount, durationMs)
	r.active.Store(true)

	// done channel is closed after duration to release all goroutines
	done := make(chan struct{})

	for i := 0; i < threadCount; i++ {
		go func() {
			// Block on a channel — simulates a leaked goroutine consuming
			// scheduler resources.
			block := make(chan struct{})
			select {
			case <-block: // never fires
			case <-done:
			}
		}()
	}

	// Cleanup after duration
	go func() {
		time.Sleep(time.Duration(durationMs) * time.Millisecond)
		close(done)
		r.active.Store(false)
		log.Println("[chaos-sdk] Goroutine leak fault completed")
	}()
}

// ---------------------------------------------------------------------------
// 2. Memory pressure — allocate and hold []byte slices
// ---------------------------------------------------------------------------

func (r *ResourceFaultInjector) applyMemoryPressure(rule chaos.FaultRuleConfig) {
	allocationMB := chaos.Clamp(chaos.GetConfigInt(rule, "allocation_mb", 64), 1, 512)
	durationMs := chaos.Clamp(chaos.GetConfigInt(rule, "duration_ms", 30000), 1000, 60000)

	log.Printf("[chaos-sdk] Injecting memory pressure: %dMB for %dms", allocationMB, durationMs)
	r.active.Store(true)

	blocks := make([][]byte, 0, allocationMB)
	func() {
		defer func() {
			if rec := recover(); rec != nil {
				log.Printf("[chaos-sdk] Memory pressure: allocation error, holding partial (%d MB)", len(blocks))
			}
		}()
		for i := 0; i < allocationMB; i++ {
			block := make([]byte, 1024*1024) // 1 MB
			// Touch all pages to ensure real allocation
			for j := 0; j < len(block); j += 4096 {
				block[j] = byte(i & 0xff)
			}
			blocks = append(blocks, block)
		}
		log.Printf("[chaos-sdk] Memory pressure: allocated %dMB", allocationMB)
	}()

	// Hold for duration, then release
	go func() {
		time.Sleep(time.Duration(durationMs) * time.Millisecond)
		// Release references — GC will collect
		blocks = nil //nolint:ineffassign
		runtime.GC() // Hint GC
		r.active.Store(false)
		log.Println("[chaos-sdk] Memory pressure fault completed, memory released")
	}()
}

// ---------------------------------------------------------------------------
// 3. CPU burn — LockOSThread + tight math loops
// ---------------------------------------------------------------------------

func (r *ResourceFaultInjector) applyCPUBurn(rule chaos.FaultRuleConfig) {
	threadCount := chaos.Clamp(chaos.GetConfigInt(rule, "thread_count", 2), 1, 8)
	durationMs := chaos.Clamp(chaos.GetConfigInt(rule, "duration_ms", 30000), 1000, 60000)

	log.Printf("[chaos-sdk] Injecting CPU burn: %d threads for %dms", threadCount, durationMs)
	r.active.Store(true)

	deadline := time.Now().Add(time.Duration(durationMs) * time.Millisecond)
	var completed atomic.Int32

	for i := 0; i < threadCount; i++ {
		go func() {
			runtime.LockOSThread()
			defer runtime.UnlockOSThread()

			x := 1.1
			for time.Now().Before(deadline) {
				// Tight math loop — burns CPU
				for j := 0; j < 10000; j++ {
					x = math.Sin(x)*math.Cos(x) + math.Sqrt(math.Abs(x))
				}
			}
			_ = x // prevent optimisation

			if int(completed.Add(1)) == threadCount {
				r.active.Store(false)
				log.Printf("[chaos-sdk] CPU burn fault completed (%d threads)", threadCount)
			}
		}()
	}
}

// ---------------------------------------------------------------------------
// 4. GC pressure — rapid short-lived allocations
// ---------------------------------------------------------------------------

func (r *ResourceFaultInjector) applyGCPressure(rule chaos.FaultRuleConfig) {
	allocRateMBPerSec := chaos.Clamp(chaos.GetConfigInt(rule, "allocation_rate_mb_per_sec", 10), 1, 100)
	durationMs := chaos.Clamp(chaos.GetConfigInt(rule, "duration_ms", 30000), 1000, 60000)

	log.Printf("[chaos-sdk] Injecting GC pressure: %dMB/sec for %dms", allocRateMBPerSec, durationMs)
	r.active.Store(true)

	go func() {
		deadline := time.Now().Add(time.Duration(durationMs) * time.Millisecond)
		interval := time.Duration(float64(time.Second) / float64(allocRateMBPerSec))
		if interval < time.Millisecond {
			interval = time.Millisecond
		}

		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				if time.Now().After(deadline) {
					r.active.Store(false)
					log.Println("[chaos-sdk] GC pressure fault completed")
					return
				}
				// Allocate 1 MB — immediately discard so GC must collect
				garbage := make([]byte, 1024*1024)
				garbage[0] = 1
				garbage[len(garbage)-1] = 1
				_ = garbage // goes out of scope, eligible for GC
			}
		}
	}()
}

// ---------------------------------------------------------------------------
// 5. Thread deadlock — goroutines blocked on channel with no writer
// ---------------------------------------------------------------------------

func (r *ResourceFaultInjector) applyThreadDeadlock(rule chaos.FaultRuleConfig) {
	durationMs := chaos.Clamp(chaos.GetConfigInt(rule, "duration_ms", 30000), 1000, 60000)

	log.Printf("[chaos-sdk] Injecting simulated deadlock for %dms", durationMs)
	r.active.Store(true)

	// done channel released after duration
	done := make(chan struct{})

	// Create goroutines that simulate deadlock: each waits on a channel
	// that no other goroutine will ever write to.
	const deadlockedCount = 4
	for i := 0; i < deadlockedCount; i++ {
		go func(id int) {
			block := make(chan struct{})
			select {
			case <-block: // never fires — simulates deadlock
			case <-done:
			}
		}(i)
	}

	go func() {
		time.Sleep(time.Duration(durationMs) * time.Millisecond)
		close(done)
		r.active.Store(false)
		log.Println("[chaos-sdk] Simulated deadlock fault completed")
	}()
}

// ---------------------------------------------------------------------------
// 6. Disk fill — write temp files
// ---------------------------------------------------------------------------

func (r *ResourceFaultInjector) applyDiskFill(rule chaos.FaultRuleConfig) {
	allocationMB := chaos.Clamp(chaos.GetConfigInt(rule, "allocation_mb", 64), 1, 512)
	durationMs := chaos.Clamp(chaos.GetConfigInt(rule, "duration_ms", 30000), 1000, 60000)

	log.Printf("[chaos-sdk] Injecting disk fill: %dMB for %dms", allocationMB, durationMs)
	r.active.Store(true)

	tempDir := os.TempDir()
	tempFiles := make([]string, 0, allocationMB)
	block := make([]byte, 1024*1024) // 1 MB
	for i := range block {
		block[i] = 0xAA
	}

	for i := 0; i < allocationMB; i++ {
		filePath := filepath.Join(tempDir, fmt.Sprintf("chaos-disk-fill-%d.tmp", i))
		if err := os.WriteFile(filePath, block, 0644); err != nil {
			log.Printf("[chaos-sdk] Disk fill: error writing %s: %v", filePath, err)
			break
		}
		tempFiles = append(tempFiles, filePath)
	}
	log.Printf("[chaos-sdk] Disk fill: wrote %dMB of temp files", len(tempFiles))

	// Hold for duration, then clean up
	go func() {
		time.Sleep(time.Duration(durationMs) * time.Millisecond)
		for _, fp := range tempFiles {
			_ = os.Remove(fp) // ignore errors
		}
		r.active.Store(false)
		log.Printf("[chaos-sdk] Disk fill fault completed, %d temp files cleaned up", len(tempFiles))
	}()
}
